cdap-security/src/main/java/io/cdap/cdap/security/authorization/DefaultAccessEnforcer.java

```diff
@@ -351,4 +408,60 @@ private MetricsContext createEntityIdMetricsContext(EntityId entityId) {
     return metricsCollectionService == null ? new NoopMetricsContext(tags)
         : metricsCollectionService.getContext(tags);
   }
+
+  @Override
+  public void createRole(Role role) throws AccessException {
+    AuthorizationResult authorizationResult = accessControllerInstantiator.get()
+      .createRole(authenticationContext.getPrincipal(), role);
+    AuthorizationUtil.setAuthorizationDataInContext(authorizationResult);
+    AuthorizationUtil.throwIfUnauthorized(authorizationResult);
+  }
+
+  @Override
+  public void dropRole(Role role) throws AccessException {
+    AuthorizationResult authorizationResult = accessControllerInstantiator.get()
+      .dropRole(authenticationContext.getPrincipal(), role);
+    AuthorizationUtil.setAuthorizationDataInContext(authorizationResult);
+    AuthorizationUtil.throwIfUnauthorized(authorizationResult);
+  }
+
+  @Override
+  public void addRoleToPrincipal(Role role, Principal principal) throws AccessException {
+    AuthorizationResult authorizationResult = accessControllerInstantiator.get()
+      .addRoleToPrincipal(authenticationContext.getPrincipal(), role, principal);
+    AuthorizationUtil.setAuthorizationDataInContext(authorizationResult);
+    AuthorizationUtil.throwIfUnauthorized(authorizationResult);
+  }
+
+  @Override
+  public void removeRoleFromPrincipal(Role role, Principal principal) throws AccessException {
+    AuthorizationResult authorizationResult =
+      accessControllerInstantiator.get().removeRoleFromPrincipal(authenticationContext.getPrincipal(), role, principal);
+    AuthorizationUtil.setAuthorizationDataInContext(authorizationResult);
+    AuthorizationUtil.throwIfUnauthorized(authorizationResult);
+  }
+
+  @Override
+  public Set<Role> listRoles(Principal principal) throws AccessException {
+    Pair<Set<Role>, AuthorizationResult> roleResultPair =
+      accessControllerInstantiator.get().listRoles(authenticationContext.getPrincipal(), principal);
+    AuthorizationUtil.setAuthorizationDataInContext(roleResultPair.getValue());
+    return roleResultPair.getKey();
```

Comment:
You should throwIfUnauthorized in all methods as well

Reply:
And do metrics

Reply:
For metrics  , according to the usage. we need metrics context which comes from `MetricsContext metricsContext = createEntityIdMetricsContext(entity);` 

We have value for `entity` in enforcement calls , but not in these calls 

We can pass `createEntityIdMetricsContext(null)` but that will give no-op context.